packages/material-ui-unstyled/src/FormControlUnstyled/FormControlUnstyled.tsx

```diff
@@ -0,0 +1,234 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
+import FormControlUnstyledContext, { FormControlUnstyledState } from './FormControlContext';
+import appendStyleProps from '../utils/appendStyleProps';
+import classes from './formControlUnstyledClasses';
+import FormControlUnstyledProps, { NativeFormControlElement } from './FormControlUnstyledProps';
+
+function hasValue(value: unknown) {
+  return value != null && !(Array.isArray(value) && value.length === 0) && value !== '';
+}
+
+function isFilled(value: unknown, defaultValue?: unknown) {
+  return hasValue(value) || hasValue(defaultValue);
+}
+
+type NonOptionalStyleProps = 'disabled' | 'error' | 'focused' | 'required';
+
+export type FormControlUnstyledStyleProps = Omit<FormControlUnstyledProps, NonOptionalStyleProps> &
+  Required<Pick<FormControlUnstyledProps, NonOptionalStyleProps>> & {
+    filled: boolean;
+  };
+
+/**
+ * Provides context such as filled/focused/error/required for form inputs.
+ * Relying on the context provides high flexibility and ensures that the state always stays
+ * consistent across the children of the `FormControl`.
+ * This context is used by the following components:
+ *
+ * *   FormLabel
+ * *   FormHelperText
+ * *   Input
+ * *   InputLabel
+ *
+ * You can find one composition example below and more going to [the demos](https://material-ui.com/components/text-fields/#components).
+ *
+ * ```jsx
+ * <FormControl>
+ *   <InputLabel htmlFor="my-input">Email address</InputLabel>
+ *   <Input id="my-input" aria-describedby="my-helper-text" />
+ *   <FormHelperText id="my-helper-text">We'll never share your email.</FormHelperText>
+ * </FormControl>
+ * ```
+ *
+ * ⚠️ Only one `Input` can be used within a FormControl because it create visual inconsistencies.
+ * For instance, only one input can be focused at the same time, the state shouldn't be shared.
+ *
+ * Demos:
+ *
+ * - [Text Fields](https://material-ui.com/components/text-fields/)
+ *
+ * API:
+ *
+ * - [FormControlUnstyled API](https://material-ui.com/api/form-control-unstyled/)
+ */
+const FormControlUnstyled = React.forwardRef(function FormControlUnstyled(
+  props: FormControlUnstyledProps,
+  ref: React.ForwardedRef<any>,
+) {
+  const {
+    defaultValue,
+    children,
+    className,
+    component,
+    components = {},
+    componentsProps = {},
+    disabled = false,
+    error = false,
+    focused: visuallyFocused = false,
+    onChange,
+    required = false,
+    value,
+    ...other
+  } = props;
+
+  const isControlled = value !== undefined;
+  const [filled, setFilled] = React.useState(() => isFilled(value, defaultValue));
```

Comment:
`filled` shouldn't need to be a separate state since it depends on `value`. The value of `filled` can be directly derived from `value`.

Reply:
With the current implementation, it must be a separate state, as the inner input can be either controlled or uncontrolled.
But now it got to me this could be simplified in a way that even if FormControlUnstyled gets `defaultValue`, it won't propagate it to the context, but rather always set a `value`. This way `filled` could indeed be simply derived from value. Do you think this would be a better approach?

Reply:
> This way `filled` could indeed be simply derived from value. Do you think this would be a better approach?

Yep, that's the goal. The current implementation is more complicated than it actually needs to. By ensuring a consistent state we also avoid cascading updates.